File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
public class Solution {
    static class Node {
         String genre;
         int play;
         int index;
         Node(String genre, int play, int index){
             this.genre=genre;
             this.play=play;
             this.index=index;
         }
        
    }
    static class MyComparator implements Comparator<Node>{
        @Override
        public int compare(Node o1, Node o2){
            if(map.get(o1.genre) > map.get(o2.genre)){
                return -1;
            }else if(map.get(o1.genre) == map.get(o2.genre)){
                if(o1.play > o2.play){
                    return -1;
                }
            }
            return 1;
        }
        
    }
    static String temp;
    static int id,cnt;
    static Map <String,Integer> map;
    public static int[] solution(String[] genres, int[] plays) {
        map = new HashMap<>();
        List <Node> list = new ArrayList<>();
        temp = "";
        for(int i=0; i<genres.length; i++){
            map.put(genres[i],map.getOrDefault(genres[i],0)+plays[i]);
            list.add(new Node(genres[i],plays[i],i));
        }
        
        MyComparator comp = new MyComparator();
        Collections.sort( list,comp);
     
        
        
        temp = list.get(0).genre;
        cnt = 1;
        id = 0;
        int[] answer = new int[200];
        for(int i=0; i<list.size(); i++){
            if(!temp.equals(list.get(i).genre)){
                temp = list.get(i).genre;
                cnt = 1;
                answer[id]=list.get(i).index;
                id++;
                cnt++;
            }else if(cnt < 3){
                answer[id]=list.get(i).index;
                id++;
                cnt++;
               
            }
            
            
        }
        answer = Arrays.copyOfRange(answer,0,id);
       
        return answer;
    }
    
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import deque
def solution(genres, plays):
    li = []
    dic = {}
    for i,(g,p) in enumerate(zip(genres, plays)):
        if g in dic:
            dic[g]+=p
        else:
            dic[g]=p
    for i,(g,p) in enumerate(zip(genres, plays)):
        li.append((-dic[g],g,-p,i))
    li.sort()
    q = deque(li)
    
    cnt=2
    nowg = li[0][1]
    answer = []
    while q:
        now = q.popleft()
        if now[1]!=nowg:
            nowg=now[1]
            cnt=2
            
        if cnt >0:
            answer.append(now[-1])
            cnt-=1
        
            
        
        
        
    
    
            
        
        
        
    
    return answer